evm/vm/base.py

```diff
@@ -133,45 +133,48 @@ def execute_bytecode(self,
     #
     def import_block(self, block):
         self.receipts = []
-        self.configure_header(
+        self.block = self.block.copy(header=self.configure_header(
             coinbase=block.header.coinbase,
             gas_limit=block.header.gas_limit,
             timestamp=block.header.timestamp,
             extra_data=block.header.extra_data,
             mix_hash=block.header.mix_hash,
             nonce=block.header.nonce,
             uncles_hash=keccak(rlp.encode(block.uncles)),
-        )
+        ))
 
         # run all of the transactions.
         for transaction in block.transactions:
             self.apply_transaction(transaction)
 
         # transfer the list of uncles.
-        self.block.uncles = block.uncles
+        self.block = self.block.copy(
+            uncles=block.uncles,
+            transactions=block.transactions,
+            header=self.block.header,
+        )
 
         return self.mine_block()
 
     def mine_block(self, *args, **kwargs):
         """
         Mine the current block. Proxies to self.pack_block method.
         """
-        block = self.block
-        tx_root_hash, tx_kv_nodes = make_trie_root_and_nodes(block.transactions)
+        tx_root_hash, tx_kv_nodes = make_trie_root_and_nodes(self.block.transactions)
         receipt_root_hash, receipt_kv_nodes = make_trie_root_and_nodes(self.receipts)
         self.chaindb.persist_trie_data_dict(tx_kv_nodes)
         self.chaindb.persist_trie_data_dict(receipt_kv_nodes)
-        self.block.header.transaction_root = tx_root_hash
-        self.block.header.receipt_root = receipt_root_hash
 
-        self.pack_block(block, *args, **kwargs)
+        with self.block.header.build_copy() as header:
+            header.transaction_root = tx_root_hash
+            header.receipt_root = receipt_root_hash
+
+        block = self.pack_block(self.block.copy(header=header), *args, **kwargs)
```

Comment:
I don't quite like the fact that `header` is used after we leave the contextmanager... Not sure if that's an indication a contextmanager is not the right solution here or if maybe we should just move the `pack_block()` call inside the contextmanager. Thoughts?

Reply:
Ok, that feeling came from me not understanding the purpose of the context manager... it was not immediately clear to me that what it does is make the object mutable inside the context, so I guess it makes sense to leave `pack_block()` outside of it as it shouldn't mutate `header`. Still, it doesn't feel right to use `header` outside of the contextmanager, specially since it gives no indication that it's no longer mutable after you exit the context. Will try to think of alternatives, but just thought I'd write this down here in case you guys have any ideas

Reply:
Linking to https://github.com/ethereum/pyrlp/pull/77 where this API is being created.

I generally agree.  see [this comment](https://github.com/ethereum/pyrlp/pull/77#pullrequestreview-114145636) for another reason why this API is probably not quite right.

Reply:
@carver @gsalgado have a look at https://github.com/ethereum/pyrlp/pull/78